File: src/test/java/spring/graphql/modules/category/CategoryRepositoryTest.java
package spring.graphql.modules.category;

import static java.util.List.of;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.when;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.nextspring.modules.category.CategoryEntity;
import com.nextspring.modules.category.CategoryRepository;

@ExtendWith(MockitoExtension.class)
class CategoryRepositoryTest {
	@Mock
	private CategoryRepository categoryRepository;

	@BeforeEach
	void setUp() {
	}

	@Test
	public void create() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "test");

		when(categoryRepository.save(categoryEntity)).thenReturn(categoryEntity);

		CategoryEntity result = categoryRepository.save(categoryEntity);

		assertNotNull(result);
		assertEquals(categoryEntity, result);
	}

	@Test
	public void findById() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "test");

		when(categoryRepository.findById(categoryEntity.getId())).thenReturn(java.util.Optional.of(categoryEntity));

		CategoryEntity result = categoryRepository.findById(categoryEntity.getId()).orElseThrow();

		assertNotNull(result);
		assertEquals(categoryEntity, result);
	}

	@Test
	public void findAll() {
		CategoryEntity firstCategory = new CategoryEntity(null, "First Category");
		CategoryEntity secondCategory = new CategoryEntity(null, "Second Category");

		when(categoryRepository.findAll()).thenReturn(of(firstCategory, secondCategory));

		Iterable<CategoryEntity> result = categoryRepository.findAll();

		assertNotNull(result);
		result.forEach(categoryEntity -> {
			assertNotNull(categoryEntity);
			assertTrue(categoryEntity.getName().equals(firstCategory.getName())
					|| categoryEntity.getName().equals(secondCategory.getName()));
		});

		assertEquals(2L, result.spliterator().getExactSizeIfKnown());
	}

	@Test
	public void deleteById() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "test");

		when(categoryRepository.findById(categoryEntity.getId())).thenReturn(java.util.Optional.of(categoryEntity));

		CategoryEntity result = categoryRepository.findById(categoryEntity.getId()).orElseThrow();

		assertNotNull(result);
		assertEquals(categoryEntity, result);
	}

	@Test
	public void update() {
		CategoryEntity categoryEntity = new CategoryEntity(null, "test");

		when(categoryRepository.findById(categoryEntity.getId())).thenReturn(java.util.Optional.of(categoryEntity));

		CategoryEntity result = categoryRepository.findById(categoryEntity.getId()).orElseThrow();

		assertNotNull(result);
		assertEquals(categoryEntity, result);
	}
}

File: src/test/kotlin/spring/graphql/modules/category/CategoryRepositoryTest.kt
package spring.graphql.modules.category

import com.nextspring.modules.category.CategoryEntity
import com.nextspring.modules.category.CategoryRepository
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import java.util.*
import java.util.function.Consumer

@ExtendWith(MockitoExtension::class)
internal class CategoryRepositoryTest {
    @Mock
    private val categoryRepository: CategoryRepository = Mockito.mock(CategoryRepository::class.java)

    @BeforeEach
    fun setUp() {
    }

    @Test
    fun create() {
        val categoryEntity = CategoryEntity(null, "test")

        Mockito.`when`(categoryRepository.save(categoryEntity)).thenReturn(categoryEntity)

        val result = categoryRepository.save(categoryEntity)

        Assertions.assertNotNull(result)
        Assertions.assertEquals(categoryEntity, result)
    }

    @Test
    fun findById() {
        val categoryEntity = CategoryEntity(null, "test")

        Mockito.`when`(categoryRepository.findById(categoryEntity.id!!)).thenReturn(Optional.of(categoryEntity))

        val result = categoryRepository.findById(categoryEntity.id!!).orElseThrow()!!

        Assertions.assertNotNull(result)
        Assertions.assertEquals(categoryEntity, result)
    }

    @Test
    fun findAll() {
        val firstCategory = CategoryEntity(null, "First Category")
        val secondCategory = CategoryEntity(null, "Second Category")

        Mockito.`when`(categoryRepository.findAll()).thenReturn(listOf(firstCategory, secondCategory))

        val result: Iterable<CategoryEntity?> = categoryRepository.findAll()

        Assertions.assertNotNull(result)
        result.forEach(Consumer { categoryEntity: CategoryEntity? ->
            Assertions.assertNotNull(categoryEntity)
            Assertions.assertTrue(categoryEntity!!.name == firstCategory.name || categoryEntity.name == secondCategory.name)
        })

        Assertions.assertEquals(2L, result.spliterator().exactSizeIfKnown)
    }

    @Test
    fun deleteById() {
        val categoryEntity = CategoryEntity(null, "test")

        Mockito.`when`(categoryRepository.findById(categoryEntity.id!!)).thenReturn(Optional.of(categoryEntity))

        val result = categoryRepository.findById(categoryEntity.id!!).orElseThrow()!!

        Assertions.assertNotNull(result)
        Assertions.assertEquals(categoryEntity, result)
    }

    @Test
    fun update() {
        val categoryEntity = CategoryEntity(null, "test")

        Mockito.`when`(categoryRepository.findById(categoryEntity.id!!)).thenReturn(Optional.of(categoryEntity))

        val result = categoryRepository.findById(categoryEntity.id!!).orElseThrow()!!

        Assertions.assertNotNull(result)
        Assertions.assertEquals(categoryEntity, result)
    }
}